docs/source/format/Columnar.rst

```diff
@@ -0,0 +1,1266 @@
+  `Message.fbs`_.
+* The ``body``, a flat sequence of memory buffers written end-to-end
+  with appropriate padding to ensure a minimum of 8-byte alignment
+
+The data header contains the following:
+
+* The length and null count for each flattened field in the record
+  batch
+* The memory offset and length of each constituent ``Buffer`` in the
+  record batch's body
+
+Fields and buffers are flattened by a pre-order depth-first traversal
+of the fields in the record batch. For example, let's consider the
+schema ::
+
+    col1: Struct<a: Int32, b: List<item: Int64>, c: Float64>
+    col2: Utf8
+
+The flattened version of this is: ::
+
+    FieldNode 0: Struct name='col1'
+    FieldNode 1: Int32 name=a'
+    FieldNode 2: List name='b'
+    FieldNode 3: Int64 name='item'
+    FieldNode 4: Float64 name='c'
+    FieldNode 5: Utf8 name='col2'
+
+For the buffers produced, we would have the following (refer to the
+table above): ::
+
+    buffer 0: field 0 validity
+    buffer 1: field 1 validity
+    buffer 2: field 1 values
+    buffer 3: field 2 validity
+    buffer 4: field 2 offsets
+    buffer 5: field 3 validity
+    buffer 6: field 3 values
+    buffer 7: field 4 validity
+    buffer 8: field 4 values
+    buffer 9: field 5 validity
+    buffer 10: field 5 offsets
+    buffer 11: field 5 data
+
+The ``Buffer`` Flatbuffers value describes the location and size of a
+piece of memory. Generally these are interpreted relative to the
+**encapsulated message format** defined next.
+
+Encapsulated message format
+---------------------------
+
+For simple streaming and file-based serialization, we define a
+so-called "encapsulated" message format for interprocess
+communication. Such messages can be "deserialized" into in-memory
+Arrow array objects by examining only the message metadata without any
+need to copy or move any of the actual data.
+
+The encapsulated binary message format is as follows:
+
+* A 32-bit continuation indicator. The value ``0xFFFFFFFF`` indicates
+  a valid message. This component was introduced in version 0.15.0 in
+  part to address the 8-byte alignment requirement of Flatbuffers
+* A 32-bit little-endian length prefix indicating the metadata size
+* The message metadata as a Flatbuffer
+* Padding bytes to an 8-byte boundary
+* The message body, which must be a multiple of 8 bytes
+
+Schematically, we have: ::
+
+    <continuation: 0xFFFFFFFF>
+    <metadata_size: int32>
+    <metadata_flatbuffer: bytes>
+    <padding>
+    <message body>
+
+The complete serialized message must be a multiple of 8 bytes so that messages
+can be relocated between streams. Otherwise the amount of padding between the
+metadata and the message body could be non-deterministic.
+
+The ``metadata_size`` includes the size of the flatbuffer plus padding. The
+``Message`` flatbuffer includes a version number, the particular message (as a
+flatbuffer union), and the size of the message body:
+
+Currently, we support 4 types of messages:
```

Comment:
It's not obvious how this relates to the above. For example, how is a Schema serialized? Is it part of the "metadata_flatbuffer"? Or is it a "message body"?